## **Урок 6. Взаимосвязь величин. Параметрические и непараметрические показатели корреляции. Корреляционный анализ.**

#### Задача 1. Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):
#### zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
#### ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].
#### Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy
#### Полученные значения должны быть равны.
#### Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков,
#### а затем с использованием функций из библиотек numpy и pandas.

In [36]:
import numpy as np
import pandas as pd

In [37]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])

In [38]:
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

Найдите ковариацию этих двух величин с помощью элементарных действий.

Формула ковариации:

$$cov_{XY} = M[(X - M(X))(Y - M(Y))] = M(XY) - M(X)M(Y)=\overline{X \cdot Y} - \overline{X} \cdot \overline{Y}$$

In [39]:
cov = np.mean(zp * ks) - np.mean(zp) * np.mean(ks)
cov

9157.839999999997

In [40]:
# а затем с помощью функции cov из numpy
#Полученные значения должны быть равны
np.cov(zp, ks, ddof = 0)

array([[ 3494.64,  9157.84],
       [ 9157.84, 30468.89]])

Зная ковариацию и среднее квадратичное отклонение каждого из двух признаков, можно вычислить коэффициент корреляции Пирсона:

$$r_{XY} = \frac{cov_{XY}}{\sigma_{X}\sigma_{Y}}$$

In [41]:
#Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков
std_zp = np.std(zp, ddof = 0)
std_zp 

59.115480206118605

In [42]:
std_ks = np.std(ks, ddof = 0)
std_ks

174.55340157098058

In [43]:
r = cov / (std_zp * std_ks)
r

0.8874900920739158

In [44]:
#а затем с использованием функций из библиотек numpy и pandas
np.corrcoef(zp, ks)

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

In [45]:
zp_, ks_ = pd.Series(zp), pd.Series(ks)

In [46]:
cov_zp_ks = zp_.cov(ks_, ddof=0)
cov_zp_ks

9157.840000000002

In [47]:
r_zp_ks = zp_.corr(ks_)
r_zp_ks

0.8874900920739163

#### Задача 2. Измерены значения IQ выборки студентов,
####  обучающихся в местных технических вузах:
####  131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
####  Известно, что в генеральной совокупности IQ распределен нормально.
#### Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [48]:
iq = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])
n= len(iq)
n

10

Так как неизвестна $\sigma$ генеральной совокупности, воспользуемся формулой нахождения доверительного интервала (по критерию Стьюдента):

$$\bar X \pm t_{\alpha/2} * \frac{\sigma_n}{\sqrt{n}}$$

In [49]:
t = 2.262
mu = np.mean(iq)
mu

118.1

In [50]:
sigma_iq = np.std(iq, ddof=1)
sigma_iq

10.54566788359614

In [51]:
a1 = mu + t * sigma_iq / np.sqrt(n)
a1

125.64339223691834

In [52]:
a2 = mu - t * sigma_iq / np.sqrt(n)
a2

110.55660776308164

#### Ответ: [110.55660776308164; 125.64339223691834].

#### Задача 3. Известно, что рост футболистов в сборной распределен нормально
#### с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27,
#### среднее выборочное составляет 174.2. Найдите доверительный интервал для математического
#### ожидания с надежностью 0.95.

Так как известна $\sigma$ генеральной совокупности, воспользуемся формулой нахождения доверительного интервала:

$$\bar X \pm Z_{\alpha/2} * \frac{\sigma}{\sqrt{n}}$$

In [53]:
sigma = 25 ** 0.5
sigma 

5.0

In [54]:
n = 27
M = 174.2
Z = -1.96

In [55]:
a1 = M + Z * sigma / np.sqrt(n)
a1

172.31398912064722

In [56]:
a2 = M - Z * sigma / np.sqrt(n)
a2

176.08601087935276

#### Ответ: [172.31398912064722; 176.08601087935276].